services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/immutable/ImmutableIntrospector.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb.internal.immutable;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbIgnore;
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbImmutable;
+
+@SdkInternalApi
+public class ImmutableIntrospector {
+    private static final ImmutableIntrospector INSTANCE = new ImmutableIntrospector();
+    private static final String BUILD_METHOD = "build";
+    private static final String BUILDER_METHOD = "builder";
+    private static final String GET_PREFIX = "get";
+    private static final String IS_PREFIX = "is";
+    private static final String SET_PREFIX = "set";
+
+    // Methods from Object are commonly overridden and confuse the mapper, automatically exclude any method with a name
+    // that matches a method defined on Object.
+    private static final Set<String> NAMES_TO_EXCLUDE =
+        Collections.unmodifiableSet(Arrays.stream(Object.class.getMethods())
+                                          .map(Method::getName)
+                                          .collect(Collectors.toSet()));
```

Comment:
I was trying not to synchronize as the static class loader path is guaranteed to be thread synchronized. Do you think it would be preferable to do a synchronized double-null-check type lazy initialization here?

Reply:
Since it's not a big deal if we double-initialize, doing a synchronized double-null-check lazy initialization wouldn't even need to be on a volatile variable. It would probably be the best of both worlds.

Reply:
Okay, will do that and make it non-volatile.

Reply:
As discussed offline and due to spotbug's protests we'll make this volatile after all.